File: web-app/frontend/lib/api/alerts.ts
/**
 * /alerts API helpers — server-side read + browser-side acknowledge mutate.
 *
 * Read path:
 *   `fetchAlerts(accessToken, params?)` runs in a server component / route
 *   handler. Needs the user's JWT from getServerSession().accessToken.
 *
 * Write path (Day 31 overlay):
 *   `updateAlertAck(anomaly_id, body)` runs in the browser. It calls our own
 *   Next.js proxy /api/alerts/{anomaly_id} which forwards the session cookie to
 *   FastAPI as Bearer. This only writes the Postgres operational state
 *   (acknowledged / dismissed / new) — Fabric is_resolved stays pipeline-owned.
 *
 * Types stay in lockstep with backend app/schemas/alerts.py.
 */

// ----------------------------------------------------------------------------
// Types
// ----------------------------------------------------------------------------

export type AlertSeverity = "CRITICAL" | "HIGH" | "MEDIUM" | "LOW"
export type AckStatus = "new" | "acknowledged" | "dismissed"
export type AlertResolution = "unresolved" | "resolved" | "all"

export type AlertItem = {
  anomaly_id: string | null
  fabric_building_id: string
  building_name: string

  anomaly_type: string | null
  severity: string | null // CRITICAL / HIGH / MEDIUM / LOW (uppercase)
  detected_at: string | null // ISO datetime
  is_resolved: boolean

  // Unit-neutral — an anomaly may be kWh, CO2 ppm, °C, or a ratio.
  metric_value: number | null
  threshold_value: number | null
  deviation_pct: number | null

  description: string | null
  recommended_action: string | null

  // Day 31 acknowledge overlay (Postgres).
  ack_status: AckStatus
  acknowledged_at: string | null
  ack_notes: string | null

  can_manage: boolean
}

export type AlertSeverityCounts = {
  critical: number
  high: number
  medium: number
  low: number
  total: number
  unresolved_total: number
  unresolved_critical: number
  unresolved_high: number
  unresolved_medium: number
  unresolved_low: number
  // Unhandled = unresolved AND not acknowledged/dismissed (triage queue + badge).
  unhandled_total: number
  unhandled_critical: number
  unhandled_high: number
  // Operational overlay tallies.
  acknowledged: number
  dismissed: number
}

export type AlertsResponse = {
  alerts: AlertItem[]
  severity_counts: AlertSeverityCounts
}

type FetchResult<T> = { ok: true; data: T } | { ok: false; error: string }

// ----------------------------------------------------------------------------
// Server-side: GET /alerts
// ----------------------------------------------------------------------------

function getBackendUrl(): string | null {
  return process.env.BACKEND_URL || null
}

export type FetchAlertsParams = {
  severity?: AlertSeverity | "all"
  building_id?: string
  unresolved_only?: boolean
  resolution?: AlertResolution
  limit?: number
}

export async function fetchAlerts(
  accessToken: string,
  params: FetchAlertsParams = {}
): Promise<FetchResult<AlertsResponse>> {
  const backendUrl = getBackendUrl()
  if (!backendUrl) {
    console.error("[fetchAlerts] BACKEND_URL not configured")
    return { ok: false, error: "Server misconfigured" }
  }
  if (!accessToken) return { ok: false, error: "Missing access token" }

  const qs = alertsQuery(params)
  const url = `${backendUrl}/alerts${qs ? `?${qs}` : ""}`

  try {
    const res = await fetch(url, {
      method: "GET",
      headers: {
        Authorization: `Bearer ${accessToken}`,
        Accept: "application/json",
      },
      cache: "no-store",
    })
    if (!res.ok) {
      const body = await res.text().catch(() => "<no body>")
      if (res.status === 503 && body.includes("fabric_unavailable")) {
        console.warn(`[fetchAlerts] Fabric data temporarily unavailable`)
        return { ok: false, error: "fabric_unavailable" }
      }
      console.error(`[fetchAlerts] Backend ${res.status}: ${body.slice(0, 200)}`)
      return { ok: false, error: `Backend ${res.status}` }
    }
    const data = (await res.json()) as AlertsResponse
    return { ok: true, data }
  } catch (err) {
    console.error("[fetchAlerts] Network error:", err)
    return { ok: false, error: "Network error" }
  }
}

// ----------------------------------------------------------------------------
// Browser-side: PATCH /alerts/{anomaly_id} via Next.js proxy
// ----------------------------------------------------------------------------

export type AlertAckUpdateRequest = {
  ack_status: AckStatus
  building_id: string
  notes?: string | null
}

export type AlertAckUpdateResponse = {
  anomaly_id: string
  ack_status: AckStatus
  acknowledged_at: string | null
}

export async function updateAlertAck(
  anomaly_id: string,
  body: AlertAckUpdateRequest
): Promise<FetchResult<AlertAckUpdateResponse>> {
  try {
    const res = await fetch(`/api/alerts/${encodeURIComponent(anomaly_id)}`, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    })
    if (!res.ok) {
      const text = await res.text().catch(() => "")
      return { ok: false, error: `Update failed (${res.status}): ${text}` }
    }
    const data = (await res.json()) as AlertAckUpdateResponse
    return { ok: true, data }
  } catch (err) {
    return {
      ok: false,
      error: err instanceof Error ? err.message : String(err),
    }
  }
}

// ----------------------------------------------------------------------------
// Shared query builder + browser-side read (client-driven filter re-fetch)
// ----------------------------------------------------------------------------

/** Build the /alerts query string shared by the server + browser reads. */
function alertsQuery(params: FetchAlertsParams): string {
  const search = new URLSearchParams()
  if (params.severity) search.set("severity", params.severity)
  if (params.building_id) search.set("building_id", params.building_id)
  if (params.unresolved_only) search.set("unresolved_only", "true")
  if (params.resolution) search.set("resolution", params.resolution)
  if (params.limit != null) search.set("limit", String(params.limit))
  return search.toString()
}

/**
 * Browser-side GET /alerts through our Next.js proxy (/api/alerts). Used by
 * AlertsShell to re-fetch when the user changes the severity or resolution
 * filter, so the table + counts come from the server (accurate under the row
 * cap) instead of slicing a stale client-loaded set. The proxy injects the
 * session JWT, so no token is needed here.
 */
export async function fetchAlertsViaProxy(
  params: FetchAlertsParams = {}
): Promise<FetchResult<AlertsResponse>> {
  const qs = alertsQuery(params)
  try {
    const res = await fetch(`/api/alerts${qs ? `?${qs}` : ""}`, {
      method: "GET",
      headers: { Accept: "application/json" },
      cache: "no-store",
    })
    if (!res.ok) {
      const body = await res.text().catch(() => "<no body>")
      if (res.status === 503 && body.includes("fabric_unavailable")) {
        console.warn("[fetchAlertsViaProxy] Fabric data temporarily unavailable")
        return { ok: false, error: "fabric_unavailable" }
      }
      console.error(`[fetchAlertsViaProxy] ${res.status}: ${body.slice(0, 200)}`)
      return { ok: false, error: `Request failed (${res.status})` }
    }
    const data = (await res.json()) as AlertsResponse
    return { ok: true, data }
  } catch (err) {
    console.error("[fetchAlertsViaProxy] Network error:", err)
    return { ok: false, error: "Network error" }
  }
}

File: web-app/backend/app/schemas/alerts.py
"""Pydantic models for /alerts endpoints.

/alerts is a portfolio-wide monitoring surface over Fabric `gold_anomaly_log`,
collapsed into ongoing *issues* (one row per building+anomaly_type) instead of
one row per daily occurrence -- see services/alerts_data.py for the rationale.

Two independent state dimensions:
  * Fabric `is_resolved`  -> analytical: did the data return to normal
    (pipeline-driven, read-only here). An issue is "open" when ANY occurrence is
    still unresolved.
  * Postgres `ack_status` -> operational: has a human acknowledged / dismissed
    it (Day 31 overlay -> alert_status table). 'new' = no row yet.

The two are NOT redundant. Field naming for the reading is unit-neutral
(metric_value / threshold_value) because an anomaly may be kWh, CO2 ppm, °C, or
a ratio -- we must not mislabel a CO2 spike as kWh. Severity values are
UPPERCASE in the source table (CRITICAL/HIGH/MEDIUM/LOW -- Day 15 finding).
"""
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

AckStatus = Literal["new", "acknowledged", "dismissed"]
"""Operational overlay state (Postgres alert_status). 'new' = no row yet."""


class AlertItem(BaseModel):
    """One row on the /alerts table -- an ongoing issue (grouped occurrences)."""

    # gold_anomaly_log has a real surrogate key (unlike gold_recommendations,
    # which forced the synthetic "building|rank" id used by /actions). After
    # grouping, this is the representative (worst+latest) occurrence's id, which
    # is also what the acknowledge overlay is keyed on.
    anomaly_id: str | None = Field(
        default=None, description="Representative gold_anomaly_log.anomaly_id."
    )

    fabric_building_id: str
    building_name: str

    anomaly_type: str | None = Field(
        default=None, description="e.g. SOLAR_PR_DROP, CONSUMPTION_SPIKE, CO2_HIGH."
    )
    severity: str | None = Field(
        default=None, description="CRITICAL / HIGH / MEDIUM / LOW (uppercase)."
    )
    detected_at: datetime | None = None
    is_resolved: bool = Field(
        default=False, description="From Fabric is_resolved bit (pipeline-driven)."
    )

    # ---- Issue rollup (grouping: one row per ongoing (building, type) issue) ----
    # gold_anomaly_log keys each anomaly by (building, type, DATE), so a chronic
    # problem spawns a fresh row every day. The service collapses those into one
    # "issue": the worst+latest occurrence is the representative, and these three
    # fields summarise the rest. occurrence_count == 1 means a one-off event.
    occurrence_count: int = Field(
        default=1, description="How many daily occurrences this ongoing issue spans."
    )
    first_detected_at: datetime | None = Field(
        default=None, description="Earliest occurrence in the issue (active since)."
    )
    last_detected_at: datetime | None = Field(
        default=None, description="Most recent occurrence in the issue."
    )

    # Unit-neutral on purpose -- see module docstring.
    metric_value: float | None = Field(
        default=None, description="The observed value that tripped the rule."
    )
    threshold_value: float | None = Field(
        default=None, description="The expected / baseline value for the rule."
    )
    deviation_pct: float | None = Field(
        default=None,
        description="(metric - threshold) / threshold * 100; None when threshold is 0/null.",
    )

    description: str | None = Field(default=None, description="description_en from Fabric.")
    recommended_action: str | None = Field(
        default=None, description="recommended_action_en from Fabric."
    )

    # ---- Postgres acknowledge overlay (Day 31) ----
    ack_status: AckStatus = Field(
        default="new",
        description="Operational state: new (no row) / acknowledged / dismissed.",
    )
    acknowledged_at: datetime | None = Field(
        default=None, description="When last acknowledged/dismissed; None when new."
    )
    ack_notes: str | None = Field(default=None, description="Free-text triage notes.")

    can_manage: bool = Field(
        default=False,
        description=(
            "True if the caller may change this alert's triage status (own org or "
            "full-manage partner). False = read-only (e.g. sample buildings)."
        ),
    )


class AlertSeverityCounts(BaseModel):
    """Severity distribution across the visible *issues* -- drives chips, cards,
    nav badge. Computed over the grouped issue set (NOT raw occurrences and NOT
    the row cap), so the numbers and the table stay consistent.
    """

    # Totals across resolved + unresolved issues.
    critical: int = 0
    high: int = 0
    medium: int = 0
    low: int = 0
    total: int = 0

    # Open-issue breakdown (>= 1 unresolved occurrence).
    unresolved_total: int = 0
    unresolved_critical: int = 0
    unresolved_high: int = 0
    unresolved_medium: int = 0
    unresolved_low: int = 0

    # Unhandled = open issue AND representative not acknowledged/dismissed.
    unhandled_total: int = 0
    unhandled_critical: int = 0
    unhandled_high: int = 0

    # Operational overlay tallies (Postgres).
    acknowledged: int = 0
    dismissed: int = 0


class AlertsResponse(BaseModel):
    """The /alerts GET payload."""

    alerts: list[AlertItem]
    severity_counts: AlertSeverityCounts


class AlertAckUpdateRequest(BaseModel):
    """Body for PATCH /alerts/{anomaly_id}.

    `building_id` is required for authorization + org resolution (the anomaly_id
    alone does not carry the building, unlike the /actions synthetic id).
    """

    ack_status: AckStatus
    building_id: str = Field(description="fabric_building_id the anomaly belongs to.")
    notes: str | None = Field(default=None, max_length=2000)


class AlertAckUpdateResponse(BaseModel):
    """Returned after a successful PATCH."""

    anomaly_id: str
    ack_status: AckStatus
    acknowledged_at: datetime | None = None
